### ISSUE TITLE
Defensive issubclass check inserts type[Unknown] to other branches

### ISSUE THREAD

**Issue by maxfischer2781** [microsoft/pyright#9967]
**Describe the bug**
Given a union `obj: type[A] | …`, where the second option is not a type, PyRight considers it sufficient to check `if isinstance(obj, type):` to narrow `obj: type[A]` for the branch. Other branches then only have the `obj: ...` remaining.

However, for runtime diagnostics one will often also check `… and issubclass(obj, A):` to reject wrong arguments. PyRight doesn't just ignore this redundant check (or warn about it being redundant) but instead considers `obj: type[Unknown] | …` as a possibility in other branches. This then erroneously causes these other branches to fail as they do not handle any `obj: type[...]` cases.

This appears wrong (or at the very least an issue when the mismatch is reported) as *a)* a new option `type[Unknown]` is added to the options of `obj`  even though it never included that and *b)* the redundant `issubclass` triggers this even though it should not affect type narrowing further.

**Code or Screenshots**
This is a simplified version of the real code that triggered an error report after a recent PyRight update. Removing the `…  and issubclass(arg, int)` check also removes the error.

```python
from typing import Callable, Any
from asyncio import iscoroutinefunction

def foo(arg: type[int] | Callable[[], Any]):
    if isinstance(arg, type) and issubclass(arg, int):
        print("it's a type[A]! 🎉")
    elif iscoroutinefunction(arg):  # error: Argument type is partially unknown
        print("it's a Bee! 🎉")     # Argument type is "(() -> int) | type[Unknown]"
    ...
```

There seems to be some dependence on the second part of the union, though I could not narrow it down much. For example, if the second part is just `str` then PyRight only complains about the `Unnecessary issubclass call; "type[int]" is always a subclass of "int"`. Note that my real code uses a callable option that is incompatible with the type option; I've used `Any` for simplicity here.

**VS Code extension or command-line**
Command-line and VS Code extension, both with strict mode.

This error occurs since pyright-1.1.392. It is not reported in pyright-1.1.391.

**Comment by erictraut** [comment]
Pyright doesn't produce any errors by default with this code. The error shows up only in strict model, as you've noted.

Let's simplify the code sample to see if we can determine why pyright's narrowing logic produces these results. I'll eliminate the `type[int]` from the union and invert the sense of the `if` conditional expression so we can understand what's happening in the `else` branch.

```python
def foo(arg: Callable[[], Any]):
    if not isinstance(arg, type) or not issubclass(arg, int):
        reveal_type(arg)
```

In the simplified example, you can see that the `if` branch will be taken under one of two conditions:
1. `arg` is not an instance of `type` or
2. `arg` is an instance of `type` but is not a subclass of `int`

That means the correct narrowed type of `arg` within this code block is `(() -> Any) |  type[Unknown]`.

Prior to version 1.1.392, pyright narrowed this type to `(() -> Any) | type[object]`. It was changed in response to [this bug report](https://github.com/microsoft/pyright/issues/9656). So this was an intentional change, not a regression.

Using `isinstance` and `issubclass` with generics will sometimes necessarily result in partially-`Unknown` types. This can unfortunately trigger strict-mode checks, but these strict-mode checks are designed to tell you about potential typing holes like this.

I don't think there's anything actionable here, so I'm going to close this as "won't fix".

**Comment by maxfischer2781** [comment]
Thanks for the explanation. So this is actually from using `Callable` since types are callable and thus an option for the `arg` type?
